keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ResourceEvent.kt

```diff
@@ -367,23 +374,37 @@ data class ResourceCheckUnresolvable(
     )
 }
 
+enum class ResourceCheckErrorOrigin {
+  USER, SYSTEM;
+  companion object {
+    fun fromException(e: SpinnakerException) =
+      when (e) {
+        is UserException -> USER
+        is SystemException -> SYSTEM
+        else -> throw SystemException("All keel exceptions should inherit from UserException or SystemException. This is a bug.")
+      }
+  }
+}
+
 data class ResourceCheckError(
   override val kind: ResourceKind,
   override val id: String,
   override val application: String,
   override val timestamp: Instant,
-  val exceptionType: Class<out Throwable>,
-  val exceptionMessage: String?
-) : ResourceCheckResult() {
+  val exceptionType: Class<out SpinnakerException>,
+  val exceptionMessage: String?,
+  val origin: ResourceCheckErrorOrigin
```

Comment:
The purpose of `origin` is to give a really simple hint to the UI as to what the original cause of the problem was. Was it something the user did wrong? Or something keel did wrong/couldn't do? Hopefully that helps render better error messages on the UI.